chore(docs): Combine Resources and Providers pages (#1092)
hashicorp/terraform-cdk: website/docs/cdktf/concepts/fundamentals/providers.html.md

---
layout: "docs"
page_title: "Providers and Resources"
sidebar_current: "cdktf"

> description: "Providers allow Terraform to communicate with external APIs. Learn to define providers in a CDK for Terraform application."

Reviewer: ```suggestion
description: "Providers allow Terraform to communicate with external APIs. Learn to define providers and resources in a CDK for Terraform application."
```

---

# Providers and Resources

A [provider](https://www.terraform.io/docs/language/providers/index.html) is any external API that has a plugin for Terraform. Provider plugins like the [AWS provider](https://registry.terraform.io/providers/hashicorp/aws/latest)
or the [cloud-init provider](https://registry.terraform.io/providers/hashicorp/cloudinit/latest/docs) act as a translation layer that allows Terraform to communicate with many different cloud providers, databases, and services.

![diagram: How Terraform uses plugins](images/terraform-plugin-overview.png)

Terraform uses providers to provision [resources](https://www.terraform.io/docs/language/resources/index.html), which describe one or more infrastructure objects like virtual networks and compute instances. Each provider on the [Terraform Registry](https://registry.terraform.io/) has documentation detailing available resources and their configuration options.

> In your CDK for Terraform Application, you will use your preferred programming language to define the resources you want Terraform to manage on one or more providers. This page provides details about how to use providers and resources in your application and how to use [escape hatches](#escape-hatch) to change resource behavior when necessary.

Reviewer: ```suggestion
In your CDK for Terraform (CDKTF) application, you will use your preferred programming language to define the resources you want Terraform to manage on one or more providers. This page provides details about how to use providers and resources in your application and how to use [escape hatches](#escape-hatch) to change resource behavior when necessary.
```

## Providers

> You can use both local providers and providers from the Terraform Registry in your application. CDKTF generates the required code bindings from the providers you define in [`cdktf.json`](/docs/cdktf/create-and-deploy/configuration.html). This allows you to define resources for that provider in your preferred programming language.

Reviewer: ```suggestion
You can install pre-built providers, import providers from the Terraform Registry, or reference local providers to define resources for your application. CDKTF generates the required code bindings from the providers you define in [`cdktf.json`](/docs/cdktf/create-and-deploy/configuration.html). This allows you to define resources for that provider in your preferred programming language.
```

> ### Pre-built Providers

Reviewer: ```suggestion
### Install Pre-built Providers
```

Adding a verb to be consistent w/other headings (Import Providers, Use a Local Provider)

### Define resources

Resource definitions and properties vary depending on the type of resource and the provider. Consult your provider's documentation for a full list of available resources and their configuration options.

The TypeScript example below defines a [DynamoDB table](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/dynamodb_table) resource on the AWS provider.

``` typescript
export class HelloTerra extends TerraformStack {
  constructor(scope: Construct, id: string) {
    super(scope, id);

    new AwsProvider(this, "aws", {
      region: "eu-central-1",
    });

    const region = new DataAwsRegion(this, "region");

    const table = new DynamodbTable(this, "Hello", {
      name: `my-first-table-${region.name}`,
      hashKey: "temp",
      attribute: [{ name: "id", type: "S" }],
      billingMode: "PAY_PER_REQUEST",
    });

```

The [Examples](/docs/cdktf/examples.html) page contains multiple example projects for every supported programming language. Add something about how to define resources.

### Import Existing Resources

> TODO: Add something about how you import existing resources. Please also provide an example :-)

Reviewer 2: There's no explicit import support in cdktf, but this can be done by using TF Core. However, this depends on which backend is used and is a bit fiddly. Do we want to put it there anyway?

Author: okay - thank you for the update! Let's omit then for now :-) If there's no explicit support, then I don't think we should talk about it.

> TODO: Add something about how you import existing resources. Please also provide an example :-)

Author: ```suggestion
```
